### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by ryangjchandler**
Opening this up as a draft initially just so you guys can track the progress. 

The basic idea is to migrate to **microbundle** for bundling the magic properties, write a decent test suite with **Jest** and move some bits around like utility functions and move to an `Object.start()` method instead of a generic function.

**Note**: These changes also require a repository name change to `magic-helpers` instead of `alpine-magic-helpers`. I think this is fine though since it's assumed that it is related to Alpine..

### CODE REVIEW

**Comment by ryangjchandler** [comment]
Microbundle supports multiple entry files through the `source` property inside of `package.json`, so whenever a new magic property is added, it would need to be added to that array.

Alternatively, we could invoke `microbundle` from inside a custom Node script that takes all of the files from that directory, but that seems a bit overkill at the moment.

**Comment by ryangjchandler** [comment]
I've written some tests for the `$truncate` magic method since it's the simplest one, @KevinBatdorf I think I've covered all of the important scenarios.

**Comment by KevinBatdorf** [comment]
I'm actually going to be pretty busy this week but I took a quick look and the refactoring and tests look good. 

Just noticed something else I think we should implement. Doesn't feel right having punctuation at the end of a truncated string. Maybe remove that but give the option to enable it. Not for this PR obviously but just a thought. 

`.toEqual('Lorem ipsum dolor sit amet,…')`

Another thought, maybe we set up a project board where we drop ideas and little tweaks like this?

**Comment by SimoTod** [comment]
I like the idea of having a project board with upcoming tasks. If we trim out punctuation, I propose we also remove spaces (I find it 'Lorem ipsum dolor sit amet …' weird)

**Comment by ryangjchandler** [comment]
> I like the idea of having a project board with upcoming tasks. If we trim out punctuation, I propose we also remove spaces (I find it 'Lorem ipsum dolor sit amet …' weird)

Yeah, I agree. I also had a weird case where `,` was being included in the trimmed string, so it wasn't really truncating based on words. I wonder if we can remove any trailing commas, full-stops too.

**Comment by KevinBatdorf** [comment]
Do you think we should use `Component` in place of `AlpineComponentMagicMethod`, etc? 

Also, what do you think about adding a changelog file?

**Comment by ryangjchandler** [comment]
> Do you think we should use `Component` in place of `AlpineComponentMagicMethod`, etc?
> 
> Also, what do you think about adding a changelog file?

Yeah, we could use `Component` instead if you wanted. I'll try to get some time to finish this PR off later on, but I can update it to that yeah.

Changelog makes sense. I think we need to stick to a format though, perhaps keepachangelog.com?

**Comment by KevinBatdorf** [comment]
That format looks good. Maybe preface the component name too? 

```
## [1.0.0] - 2017-06-20
### Added
- [$truncate] Adds option for ....
### Fixed
- [$parent] Adds fix for ...etc etc

```

**Comment by ryangjchandler** [comment]
> That format looks good. Maybe preface the component name too?
> 
> ```
> ## [1.0.0] - 2017-06-20
> ### Added
> - [$truncate] Adds option for ....
> ### Fixed
> - [$parent] Adds fix for ...etc etc
> ```

Yeah, I like the idea of that.

**Comment by ryangjchandler** [comment]
@SimoTod feeling risky with all those force pushes are we ;)

**Comment by SimoTod** [comment]
> @SimoTod feeling risky with all those force pushes are we ;)

why? It's to avoid a tons of wip commit.

**Comment by ryangjchandler** [comment]
> > @SimoTod feeling risky with all those force pushes are we ;)
> 
> why? It's to avoid a ton of wip commits.
> I tend to rebase/amend the last commit if relevant and force push it.

It was a bit of a joke, not being serious, you can commit however you want lol

**Comment by KevinBatdorf** [comment]
Should the build files be removed and just let the GitHub action take care of it? 

Also, should we leave the minifying to jsdelivr to handle? Or how does that work (I'm not really sure)?

The funding file can be updated to include everyone or remove it all together. It's not my project anymore so no sense having me alone in there.

**Comment by ryangjchandler** [comment]
> > @SimoTod feeling risky with all those force pushes are we ;)
> 
> why? It's to avoid a ton of wip commits.
> I tend to rebase/amend the last commit if relevant and force push it.

Also, for what it's worth, I try to only commit when I can describe the changes that I have made. The only time I really use `wip` commits is if it's time sensitive or I need to go somewhere or do something else (which recently is most of the time having 2 jobs lol)

File: src/utils.js
Comment (HugoDF) [inline src/utils.js:1]: maybe not for this PR, but there could also be a check that "Alpine version > version that added `addMagicProperty`"

File: src/truncate.js
Comment (HugoDF) [inline src/truncate.js:8]: maybe beyond the scope of this PR, but are we really using a variable number of parameters?

Pretty sure it's always `(string, options)`? would make the code a bit more readable as well

File: .github/workflows/run-tests.yml
Comment (HugoDF) [inline .github/workflows/run-tests.yml:4]: does this work? looks like it might give a YAML syntax error

**Review by HugoDF — DISMISSED** [review]
Nice one, couple of comments but they're probably beyond the scope here.

File: .github/workflows/run-tests.yml
Comment (SimoTod) [reply .github/workflows/run-tests.yml]: I think you are right. Fixed

**Review by HugoDF — APPROVED** [review]

File: src/truncate.js
Comment (KevinBatdorf) [reply src/truncate.js:8]: Just saw this. It was originally designed with an optional 3rd param to override the "..." that's appended. But we could make it `(string, options)` though and keep that format throughout all the helpers.

File: src/utils.js
Comment (HugoDF) [reply src/utils.js:1]: https://github.com/alpine-collective/alpine-magic-helpers/issues/15

File: src/truncate.js
Comment (HugoDF) [reply src/truncate.js:8]: https://github.com/alpine-collective/alpine-magic-helpers/issues/16

**Review by HugoDF — APPROVED** [review]

**Comment by KevinBatdorf** [comment]
Added some additional tests for the edge cases that had recent bug fixes.

My build seems to have changed the spacing of the build files. Is that something we should fix in the editor config of somewhere?

**Comment by SimoTod** [comment]
> My build seems to have changed the spacing of the build files. Is that something we should fix in the editor config of somewhere?

It's correct. It should have 4 spaces now. I think I didn't rebuild dist after merging the version check PR.
We need to fix the conflict in component.js... again :(

**Comment by KevinBatdorf** [comment]
> We need to fix the conflict in component.js... again :(

I'll do that now

**Comment by KevinBatdorf** [comment]
Looks like the other PR had `target[key] != null` instead of `target[key] !== null` so that was conflicting.

I think we're good to merge now if you agree. Will the GH action update the version numbers automatically after we tag?

**Review by SimoTod — APPROVED** [review]
looks good to me

**Comment by SimoTod** [comment]
> Will the GH action update the version numbers automatically after we tag?

yeah, when you tag a new version, it should update package.json and push to npm

**Comment by KevinBatdorf** [comment]
@SimoTod Could it update the readme somehow? Otherwise we should probably update that here then before merging

**Comment by SimoTod** [comment]
The cdn links? I'd like the github action being responsible for updating the readme. We can merge it master now, but I think we don't need to tag a new release until we add a new helper or we fix a bug. There won't be any benefit at the minute.

**Comment by KevinBatdorf** [comment]
Merged! Great job everyone. I think this puts the project in a much better place now.

**Comment by HugoDF** [comment]
> @SimoTod Could it update the readme somehow? Otherwise we should probably update that here then before merging

We can make this happen, a JS script that reads from the package.json and replaces versions in the README

**Comment by SimoTod** [comment]
The github action already knows the current tag and it runs on unix, I was thinking to just use sed to replace those strings so we don't need an additional script

**Comment by HugoDF** [comment]
Yeah that would work as well created an issue https://github.com/alpine-collective/alpine-magic-helpers/issues/21

**Pull request merged**
